torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -741,13 +491,206 @@ Stmt* LoopNest::lowerToStmt(Tensor* t) {
   return body;
 }
 
+class FunctionInliner : public IRMutator {
+ public:
+  FunctionInliner(Store* producer)
+      : buf_(producer->buf()), producer_(producer) {
+    for (auto* i : producer->indices()) {
+      const Var* index_var = dynamic_cast<const Var*>(i);
+      if (index_var == nullptr) {
+        throw std::logic_error("cannot inline Buf with compound indices");
+      }
+      index_vars_.insert(index_var);
+    }
+  }
+
+ protected:
+  // For the target function, insert the caller/callee pair into the replacement
+  // mapping.
+  const Expr* mutate(const FunctionCall* v) override {
+    Function* func = v->tensor()->function();
+    const Buf* buf = v->tensor()->buf();
+    if (buf != buf_) {
+      return IRMutator::mutate(v);
+    }
+
+    // TODO: Support multiple-output functions
+    if (func->func_vars().size() != 1) {
+      throw unimplemented_lowering();
+    }
+
+    std::vector<const Var*> index_vars;
+    for (size_t i = 0; i < buf->ndim(); i++) {
+      const Var* func_callee_arg = dynamic_cast<const Var*>(func->arg(i));
+      const Expr* func_caller_param = v->param(i);
+      auto iter = inline_mapping_.find(func_callee_arg);
+      if (iter != inline_mapping_.end()) {
+        throw std::runtime_error(
+            "Duplicated variables: " + func_callee_arg->name_hint());
+      }
+      inline_mapping_[func_callee_arg] = func_caller_param;
+      index_vars.push_back(func_callee_arg);
+    }
+
+    // Call the actual replacement.
+    const Expr* body = producer_->value();
+    const Expr* result = body->accept_mutator(this);
+
+    // Remove the caller/callee relationship.
+    for (auto* v : index_vars) {
+      for (auto& pair : random_bindings_) {
+        if (pair.second.erase(v)) {
+          const Expr* inlined = inline_mapping_[v];
+          for (auto* nv : VarFinder::find(inlined)) {
+            pair.second.insert(nv);
+          }
+        }
+      }
+      inline_mapping_.erase(v);
+    }
+    return result;
+  }
+
+  // Replace the target variable with the caller expressions.
+  const Expr* mutate(const Var* v) override {
+    auto iter = inline_mapping_.find(v);
+    if (iter == inline_mapping_.end()) {
+      return v;
+    } else {
+      const Expr* expr = iter->second;
+      // Continue to transform the value from the lookup table.
+      return expr->accept_mutator(this);
+    }
+  }
+
+  // Handle random intrinsics which should be cached.
+  const Expr* mutate(const Intrinsics* v) override {
+    if (!in_producer_ || v->op_type() != kRand) {
+      return IRMutator::mutate(v);
+    }
+
+    const std::string& name = buf_->name_hint();
+    Var* new_var = new Var(name, v->dtype());
+    random_bindings_[new Let(new_var, v)] = index_vars_;
+    return new_var;
+  }
+
+  // Remove the buffer write the inlined function.
+  Stmt* mutate(const Store* v) override {
+    if (v == producer_) {
+      in_producer_ = true;
+      producer_ = dynamic_cast<const Store*>(IRMutator::mutate(v));
```

Comment:
How do we guarantee that we always return `Store` in the mutator? In theory, any `Stmt` could be returned, right?

Also, since we're removing this statement anyway, what is the reason we're calling mutator on it? Is it to collect bindings for rand calls? If yes, would it be better to extract it into a separate visitor and maybe call it in the very beginning?

Reply:
The IRMutator::mutate() method will always return a Store* (even if the signature is Stmt*) - I'll add an assert in case this changes.

We're calling the mutate because we want to mutate the producer_ Stmt we'll later insert, not only collecting random calls but replacing them with the variable we create. I could do this with two mutators I suppose, do you feel that would be clearer? (keep in mind we'd need to both mutate the producer_ internally and pass the var info between the mutators).

Reply:
Feel free to keep in one mutator if you feel that makes the implementation easier - adding some comments would help though.